batch.py

```diff
@@ -0,0 +1,389 @@
+    logging.info(
+        f"Number of rare (freq<5%) and QC'd biallelic variants: {mt.count()[0]}"
+    )
+
+
+# endregion SUBSET_VARIANTS
+
+
+# region GET_GENE_SPECIFIC_VARIANTS
+
+
+def get_promoter_variants(
+    mt_path: str,  # checkpoint from function above
+    ht_path: str,
+    gene_details: dict[str, str],  # ouput of make_gene_loc_dict
+    window_size: int,
+    plink_output_prefix: str,  # 'tob_wgs_rv/pseudobulk_rv_association/plink_files/'
+):
+    """Subset hail matrix table
+
+    Input:
+    mt_path: path to already subsetted hail matrix table
+    gene_file: path to a tsv file with information on
+    the chrom, start and end location of genes (rows)
+
+    Output:
+    For retained variants, that are: 1) in promoter regions and
+    2) within 50kb up or down-stream of the gene body (or in the gene body itself)
+    (on top of all filters done above)
+
+    returns:
+    - Path to plink prefix for genotype files (plink format: .bed, .bim, .fam)
+    - Path to hail table with variant (rows) stats for downstream analyses
+    """
+    # read hail matrix table object (pre-filtered)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    gene_name = gene_details["gene_name"]
+
+    # get relevant chromosome
+    chrom = gene_details["chrom"]
+
+    # subset to window
+    # get gene body position (start and end) and build interval
+    left_boundary = max(1, int(gene_details["gene-start"]) - window_size)
+    right_boundary = min(
+        int(gene_details["gene-end"]) + window_size,
+        hl.get_reference("GRCh38").lengths[f"chr{chrom}"],
+    )
+    # get gene-specific genomic interval
+    gene_interval = f"{chrom}:{left_boundary}-{right_boundary}"
+    logging.info(f"Interval considered: {gene_interval}")  # 'chr22:23219960-23348287'
+
+    # include variants up to {window size} up- and downstream
+    mt = hl.filter_intervals(
+        mt, [hl.parse_locus_interval(gene_interval, reference_genome="GRCh38")]
+    )
+    mt_path = output_path("in_window.mt", "tmp")
+    mt = mt.checkpoint(
+        mt_path, overwrite=True
+    )  # add checkpoint to avoid repeat evaluation
+    logging.info(f"Number of variants within interval: {mt.count()[0]}")
+
+    # annotate using VEP
+    vep_ht = hl.read_table(ht_path)
+    mt = mt.annotate_rows(vep=vep_ht[mt.row_key].vep)
+
+    # filter variants found to be in promoter regions
+    mt = mt.filter_rows(
+        mt.vep.regulatory_feature_consequences["biotype"].contains("promoter")
+    )
+    promoter_path = output_path(f"{gene_name}promoter_variants.mt", "tmp")
+    mt = mt.checkpoint(promoter_path, overwrite=True)  # checkpoint
+    logging.info(
+        f"Number of rare (freq<5%) QC-passing, biallelic SNPs in promoter regions: {mt.count()[0]}"
+    )
+
+    # export this as a Hail table for downstream analysis
+    ht_path = output_path(f"{gene_name}_rare_promoter_summary.ht")
+    ht = mt.rows()
+    ht.write(ht_path)
```

Comment:
Here you're checkpointing the MT, writing the MT as plink, and also writing the rows. Hail can export to plink, and import from plink. Could you replace these 3 writes with just the write to plink? (then [import_plink](https://hail.is/docs/0.2/methods/impex.html#hail.methods.import_plink) later if you need an alternate format?)

Reply:
do you mean specifically in case i won't actually need the hail table? because i definitely will want to save those records. Also does the export to plink definitely retain all information?

Reply:
That's a good question 😅 I only saw that an import-from plink method exists, but I don't know if the MT -> Plink -> MT process is lossy.

My main thought is that you're checkpointing, then writing two objects - the checkpoint is a good idea to make both other writes faster, but if you don't explicitly need the hail data you can reduce this down to just the one write to disc (plink). I'm not really familiar with the plink output format, so it might drop all the annotations you need (in which case the hail table is probably good to write, and this section is fine)

Reply:
yea that was my assumption, so may leave it for now (I don't think the plink format is well suited for all these annotations, though I don't really know)